src/ChainSafe.Gaming/Web3/Core/LifecycleHandler.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace ChainSafe.Gaming.Web3.Core
+{
+    /// <summary>
+    /// Handles the lifecycle of <see cref="ILifecycleParticipant"/> instances.
+    /// </summary>
+    public class LifecycleHandler
+    {
+        private readonly ILifecycleParticipant[] lifecycleParticipants;
+
+        public LifecycleHandler(IEnumerable<ILifecycleParticipant> lifecycleParticipants)
+        {
+            // Arrange execution based on ExecutionOrder Attribute priority.
+            this.lifecycleParticipants = lifecycleParticipants.OrderBy(p => p.ExecutionOrder).ToArray();
+        }
+
+        /// <summary>
+        /// Starts all lifecycle participants.
+        /// </summary>
+        /// <returns>Awaitable Task.</returns>
+        public async Task StartAsync()
+        {
+            List<ILifecycleParticipant> startedParticipants = new List<ILifecycleParticipant>();
+
+            try
+            {
+                foreach (var lifecycleParticipant in lifecycleParticipants)
+                {
+                    await lifecycleParticipant.WillStartAsync();
+
+                    startedParticipants.Add(lifecycleParticipant);
+                }
+            }
+            catch (Exception)
+            {
+                // If an exception was thrown, dispose of all initialized participants.
+                foreach (var lifecycleParticipant in startedParticipants)
+                {
+                    await lifecycleParticipant.WillStopAsync();
+                }
+
+                // Rethrow the exception.
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Stops all lifecycle participants.
+        /// </summary>
+        /// <returns>Awaitable Task.</returns>
+        public async Task StopAsync()
+        {
+            List<Exception> exceptions = new List<Exception>();
+
+            List<ILifecycleParticipant> stoppedParticipants = new List<ILifecycleParticipant>();
+
+            await StopAllParticipantsAsync();
+
+            if (exceptions.Count != 0)
+            {
+                throw new AggregateException(exceptions);
+            }
+
+            async Task StopAllParticipantsAsync()
+            {
+                try
+                {
+                    foreach (var participant in lifecycleParticipants.Where(p => !stoppedParticipants.Contains(p)))
```

Comment:
You could simply create a queue, initialize it with all the `lifecycleParticipants` and `while (queue.Count > 0)` them instead of using recursion. You could then also remove the `.Where(p => !stoppedParticipants.Contains(p)`.

Another important thing is that you should terminate things in the opposite direction comparing to how you initialize them. This means that if some service had to be initialized last (due to its execution order), it should terminate first. This is for example important if this service depends on initialization of another one. So if service B need to initialize after service A (A, B), it should terminate before service A (B, A), as B is dependent on A and A is not dependent on B.

Reply:
decent advice....removing dependency here is good.

Reply:
ahh good point, will do

Reply:
done